### PR TITLE
Can login with wrong credential #1085

### PR DESCRIPTION
**Pull request by roxy-dao**
#1085 

User shouldn't be able to log in with old password in api now if password has been changed in central.

Note: Still can use old password if not logged into the central server and new password has not been saved in remote

Test: 
- [ ] Change password in central
- [ ] Try to login with old pass in api while still connected to central
- [ ] Disconnect from central and try to log in with old password
- [ ] Connect back to central and save new password by logging in

### CODE REVIEW

File: service/src/apis/login_v4.rs [hunk 1]
```diff
@@ -128,13 +132,21 @@ impl LoginApiV4 {
...
-            .await?
-            .error_for_status()?;
+            .await
+            .map_err(|e| LoginV4Error::ConnectionError(e.into()))?;
 
-        let response = response.json().await?;
-        Ok(response)
+        match response.status() {
+            reqwest::StatusCode::UNAUTHORIZED => Err(LoginV4Error::Unauthorized),
+            _ => {
```
Comment (andreievg) [inline service/src/apis/login_v4.rs:143]: hmm, we did error_for_status previously, maybe it's a good idea to keep that check ?

Also don't really need to use anyhow error now can use reqwest::Error ?

Lastly kind of like reducing nesting (this is probably stylistic preference, see `if let` vs match in the example below):

https://github.com/openmsupply/remote-server/compare/%231085-Can-login-with-wrong-credential-...1085-pr-comment-1

You'll note that from_reqwest_error was created, although it looks like a perfect time to use from trait, but we've been deliberately avoiding it as it adds indirection that's hard to trace (we do use from impl but in cases where it is repeatedly used).

File: service/src/login.rs [hunk 2]
```diff
@@ -141,19 +141,28 @@ impl LoginService {
...
-                FetchUserError::ConnectionError(format!(
-                    "Failed to reach the central server to fetch data for {}: {:?}",
-                    username, err
-                ))
-            })?;
+        {
+            Ok(user_data) => user_data,
+            Err(err) => match err {
+                LoginV4Error::Unauthorized => {
+                    return Err(FetchUserError::Unauthenticated);
```
Comment (andreievg) [inline service/src/login.rs:155]: If Unauthorised, but can login with DB credentials, we would need to make sure to 'disable' user being able to login, i've left another comment on the issue.

> Note: Still can use old password if not logged into the central server and new password has not been saved in remote

I think we need to try and disable access to user in offline mode with their old password if they tried to login with their old password while there is a connection to central

> Test: 
> - [ ] Change password in central
> - [ ] Try to login with old pass in api while still connected to central
> - [ ] Disconnect from central and try to log in with old password
> - [ ] Connect back to central and save new password by logging in

We should try to automate testing as much as possible, see below (last test fails btw as per above)

https://github.com/openmsupply/remote-server/compare/1085-pr-comment-1...1085-pr-comment-2

File: service/src/apis/login_v4.rs (hunk 1 above)
Comment (roxy-dao) [reply service/src/apis/login_v4.rs:143]: error_for_status causes the old password login while on central 🤔

File: service/src/login.rs [hunk 3]
```diff
@@ -334,53 +346,169 @@ mod test {
...
+            )
+            .await;
+
+            assert!(
+                matches!(result, Err(LoginError::LoginFailure)),
+                "expected login failure, got {:#?}",
+                result
+            );
+        }
+        // If central server is not accessible after trying to login with old password, make sure old password does not work
```
Comment (andreievg) [inline service/src/login.rs:485]: One last thing please, can you add this issue in the comment please https://github.com/openmsupply/remote-server/issues/1101

File: service/src/apis/login_v4.rs [hunk 4]
```diff
@@ -128,13 +132,23 @@ impl LoginApiV4 {
...
-            .await?
-            .error_for_status()?;
+            .await
+            .map_err(|e| LoginV4Error::ConnectionError(e))?;
+
+        if reqwest::StatusCode::UNAUTHORIZED == response.status() {
+            return Err(LoginV4Error::Unauthorised);
+        }
+
+        let response = response
```
Comment (andreievg) [inline service/src/apis/login_v4.rs:145]: Btw don't need to do Ok(response) in case, that last await returns a result

**Review by andreievg — APPROVED** [review]
thanks for all of the changes.

only thing really before merging is adding the issue in the comment

**Review by bijaySussol — APPROVED** [review]
@roxy-dao : 
Nice, I tested it with those changes, This works great !
Thanks

**Pull request merged**